hack/windows/Install-OVS.ps1

```diff
@@ -187,31 +226,219 @@ function InstallOVS() {
             }
         }
     }
+}
 
-    # Install OVS kernel driver.
-    Log "Installing OVS kernel driver"
-    $VMMSStatus = $(Get-Service vmms -ErrorAction SilentlyContinue).Status
-    if (!$VMMSStatus) {
-        $VMMSStatus = "not exist"
+function CheckAndInstallVCRedists {
+    param (
+        [Parameter(Mandatory = $true)] [String]$VCRedistPath,
+        [Parameter(Mandatory = $true)] [String]$VCRedistsVersion
+    )
+    $mininalVersion = [version]$VCRedistsVersion
+    $existingVCRedists = getInstalledVcRedists
+    foreach ($redist in $existingVCRedists) {
+        $installedVersion = [version]$redist.Version
+        # VC redists files with a higher version are installed, return.
+        if ($installedVersion -ge $mininalVersion) {
+            return
+        }
     }
-    Log "Hyper-V Virtual Machine Management service status: $VMMSStatus"
-    if ($VMMSStatus -eq "Running") {
-        cmd /c "cd $OVSDriverDir && install.cmd"
-    } else {
-        cd $OVSDriverDir ; netcfg -l .\ovsext.inf -c s -i OVSExt; cd $WorkDir
+    # Install the provided VC redistributable files.
+    Get-ChildItem $VCRedistPath -Filter *.exe | ForEach-Object {
+        Start-Process -FilePath $_.FullName -Args '/install /passive /norestart' -Verb RunAs -Wait
+    }
+}
+
+function CheckAndInstallOVSDriver {
+    param (
+        [Parameter(Mandatory = $true)]
+        [String]$OVSDriverPath
+    )
+
+    $expVersion = [version]$(Get-Item $OVSDriverPath\ovsext.sys).VersionInfo.ProductVersion
+    $ovsInstalled = $(netcfg -q ovsext) -like "*is installed*"
+    $installedDrivers = getInstalledOVSDrivers
+
+    # OVSext driver with the desired version is already installed, return
+    if ($ovsInstalled -and ($installedDrivers.Length -eq 1) -and ($installedDrivers[0].DriverVersion -eq $expVersion)){
+        return
+    }
+
+    # Uninstall the existing driver which is with a different version.
+    if ($ovsInstalled) {
+        netcfg -u ovsext
+    }
+
+    # Clean up the installed ovsext drivers packages.
+    foreach ($driver in $installedDrivers) {
+        $publishdName = $driver.PublishedName
+        pnputil.exe -d $publishdName
     }
-    if (!$?) {
-        Log "Install OVS kernel driver failed, exit"
+
+    # Import OVSext driver certificate to TrustedPublisher and Root.
+    $DriverFile="$OVSDriverPath\ovsext.sys"
+    $CertificateFile = "$OVSDriverPath\package.cer"
```

Comment:
Is it reasonable to expect that this file is always included?
If someone brings their own driver, will they know that this file needs to be present?

Reply:
> Is it reasonable to expect that this file is always included?

In fact, this version of the script always doesn't expect the existence of the cert. Instead, it is trying to read the certificate from driver 'ovsext.cat' to file ` $CertificateFile`, and then enforce to import it into cert chain of trustedPublisher and root. This change is because I found that using Windows Server 2022, a signed OVSext driver  (not test-signed) still requires the import, otherwise it may fail to install in my test. This is a different behavior from Server 2019. So I adjust the functions here.

Reply:
I see, I misread the code.
So this should work in both cases (signed and test-signed)?

Reply:
> So this should work in both cases (signed and test-signed)?

Yes, it works for both signed and test-signed driver.